syncer/syncer_test.go

```diff
@@ -1031,6 +1034,279 @@ func (s *testSyncerSuite) TestCasuality(c *C) {
+			).AddRow("st", 0, "PRIMARY", 1, "id", "A", 0, null, null, null, "BTREE", "", ""))
+
+		// run sql on upstream db
+		runSQL(_case.testSQLs)
+		// mock expect sql
+		for i, expectSQL := range _case.expectSQLS {
+			mock.ExpectBegin()
+			if strings.HasPrefix(expectSQL.sql, "ALTER") {
+				mock.ExpectExec(expectSQL.sql).WillReturnResult(sqlmock.NewResult(1, int64(i)+1))
+				mock.ExpectCommit()
+				// mock get table after ddl sql exec
+				mock.ExpectQuery("SHOW COLUMNS").WillReturnRows(
+					sqlmock.NewRows([]string{"Field", "Type", "Null", "Key", "Default", "Extra"}).AddRow("id", "int", "NO", "PRI", null, "").AddRow("age", "int", "NO", "", null, "").AddRow("name", "varchar", "NO", "", null, ""))
+				mock.ExpectQuery("SHOW INDEX").WillReturnRows(
+					sqlmock.NewRows([]string{"Table", "Non_unique", "Key_name", "Seq_in_index", "Column_name",
+						"Collation", "Cardinality", "Sub_part", "Packed", "Null", "Index_type", "Comment", "Index_comment"},
+					).AddRow("st", 0, "PRIMARY", 1, "id", "A", 0, null, null, null, "BTREE", "", ""))
+			} else {
+				// change insert to replace because of safe mode
+				mock.ExpectExec(expectSQL.sql).WithArgs(expectSQL.args...).WillReturnResult(sqlmock.NewResult(1, 1))
+				mock.ExpectCommit()
+			}
+		}
+		ctx, cancel := context.WithCancel(context.Background())
+		resultCh := make(chan pb.ProcessResult)
+
+		go syncer.Process(ctx, resultCh)
+
+		go func() {
+			// sleep to ensure ddlExecInfo.Send() happen after ddlExecInfo.Renew() in Process
+			// because Renew() will generate new channel, Send() may send to old channel due to goroutine schedule
+			time.Sleep(1 * time.Second)
+			// mock permit exec ddl request from dm-master
+			req := &DDLExecItem{&pb.ExecDDLRequest{Exec: true}, make(chan error, 1)}
+			syncer.ddlExecInfo.Send(ctx, req)
+		}()
+
+		select {
+		case r := <-resultCh:
+			for _, err := range r.Errors {
+				c.Errorf("Case %d: Process Err:%s", i, err)
+			}
+			c.Assert(len(r.Errors), Equals, 0)
+		case <-time.After(2 * time.Second):
+		}
+		// wait for flush finish in Process.Run()
+		// cancel function only closed done channel asynchronously
+		// other goroutine(s.streamer.GetEvent()) received done msg then trigger flush
+		// so the flush action may not execute before we check due to goroutine schedule
+		cancel()
+
+		syncer.Close()
+
+		flushedGP := syncer.checkpoint.FlushedGlobalPoint().Pos
+		GP := syncer.checkpoint.GlobalPoint().Pos
+		c.Assert(GP, Equals, flushedGP)
+
+		// Xid Event has 31 bytes
+		// check whether last event before flushed globalPoint is Xid event
+		query := fmt.Sprintf("SHOW BINLOG EVENTS FROM %d LIMIT 1", GP-31)
+		c.Logf(query)
+		r, err := s.db.Query(query)
+		c.Assert(err, IsNil)
+
+		var pos, endLogPos uint32
+		var name, eventType string
+		var unusedVal interface{}
+		unused := &unusedVal
+
+		for r.Next() {
+			r.Scan(&name, &pos, &eventType, &unused, &endLogPos, &unused)
+		}
+
+		c.Assert(eventType, Equals, "Xid")
```

Comment:
@amyangfei here check last event before globalPoint is Xid event, please review it

Reply:
I think you misunderstand it, do we need to check checkpoint here @amyangfei

Reply:
we have got the flushed global point, which aspect do you want to check? @GregoryIan